core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package scalaz.zio.internal
+
+/**
+ * A very fast, hand-optimized stack designed just for booleans.
+ * In the common case (size < 256), achieves zero allocations.
+ */
+final class StackBool private () {
+  import StackBool.Entry
+
+  private[this] var head  = new Entry(null)
+  private[this] var _size = 0L
+
+  final def getOrElse(index: Int, b: Boolean): Boolean = {
+    var j   = index.toLong
+    var cur = head
+    while (j >= 256L && (cur.next ne null)) {
+      j -= 256L
+      cur = cur.next
+    }
+    assert(j < 256L)
+    if (cur eq null) b
+    else {
+      val mask = 1L << j
+
+      (cur.bits & mask) != 0L
+    }
+  }
+
+  final def size = _size
+
+  final def push(flag: Boolean): Unit = {
+    val index = _size & 0XFFFFFFFFFFFFFFFFL
```

Comment:
🔥

Reply:
Once a game programmer, always a game programmer. 😆

Reply:
Oh, and this is also broken (though I'm not sure why tests pass). So apparently, once a functional programmer, never a procedural programmer. 😜

Reply:
OK, wasn't incorrect so much as unnecessarily crazy. Fixed. 👍